File: Data Structure/Array or Matrix/Medium/LET_no251_Flatten 2D Vector.py
class Vector2D:

    def __init__(self, vec: List[List[int]]):
        self.i, self.j = 0, 0
        self.vec = vec

    def next(self) -> int:
        self.hasNext()
        self.j += 1
        return self.vec[self.i][self.j - 1]

    def hasNext(self) -> bool:
        while self.i < len(self.vec) and self.j == len(self.vec[self.i]):
            self.i += 1
            self.j = 0
        return self.i < len(self.vec)


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(vec)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Data Structure/Array or Matrix/Medium/LET_no251_Flatten 2D Vector.cpp
class Vector2D {
private:
    int i, j;
    vector<vector<int>> memo;
public:
    Vector2D(vector<vector<int>>& vec) {
        memo = vec;
        i = j = 0;
    }
    
    int next() {
        hasNext();
        int res = memo[i][j];
        j++;
        return res;
    }
    
    bool hasNext() {
        while (i < memo.size() && j == memo[i].size()) {
            i++;
            j = 0;
        }
        return i < memo.size();
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(vec);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */